### PR TITLE
PR regarding issue:  Add DynamicObstacles #33

### PR DESCRIPTION
**Pull request by RajGhugare19**
1) Added DynamicObstacles env

### CODE REVIEW

File: src/envs/doorkey.jl [hunk 1]
```diff
@@ -15,7 +15,8 @@ function DoorKey(;n=8, agent_start_pos=CartesianIndex(2,2), rng=Random.GLOBAL_RN
     world[EMPTY, 2:n-1, 2:n-1] .= true
     world[WALL, [1,n], 1:n] .= true
     world[WALL, 1:n, [1,n]] .= true
-    world[WALL, [1,n], 1:n] .= true
+    #This is redundant I think
```
Comment (findmyway) [inline src/envs/doorkey.jl:18]: Remove the comment?

File: src/envs/dynamicobstacles.jl [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    objects = (EMPTY, WALL, OBSTACLE, GOAL)
+    w = GridWorldBase(objects, n, n)
+
+    w[EMPTY, 2:n-1, 2:n-1] .= true
+    w[WALL, [1,n], 1:n] .= true
+    w[WALL, 1:n, [1,n]] .= true
+    w[GOAL, n-1, n-1] = true
+    w[EMPTY, n-1, n-1] = false
+
+    w[OBSTACLE,1:n,1:n] .= false
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:23]: I guess this line is not needed here? All values are initialized with false by default.

File: src/envs/dynamicobstacles.jl [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    #num obstacles is n-3 only for n=5 or n=6, it need not be a function of n otherwise
+    num_obstacle = n-3
+    
+    obs_pos_array = Array{CartesianIndex{2},1}(undef,num_obstacle)
+    
+    obstacles_placed = 0
+    while obstacles_placed < num_obstacle
+        obs_pos = CartesianIndex(rand(2:n-1), rand(2:n-1))
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:31]: Add a rng as parameter in constructor and call rand(rng, …) here to help increase reproducity.

File: src/envs/dynamicobstacles.jl [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    w[EMPTY, 2:n-1, 2:n-1] .= true
+    w[WALL, [1,n], 1:n] .= true
+    w[WALL, 1:n, [1,n]] .= true
+    w[GOAL, n-1, n-1] = true
+    w[EMPTY, n-1, n-1] = false
+
+    w[OBSTACLE,1:n,1:n] .= false
+    #num obstacles is n-3 only for n=5 or n=6, it need not be a function of n otherwise
+    num_obstacle = n-3
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:25]: I prefer to mov it in the constructor as a parameter with a default value of n-3

File: src/envs/dynamicobstacles.jl [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    dest = dir(w.agent_pos)
+    if !w.world[WALL, dest]
+        #if w.world[OBSTACLE, dest]
+        #    #end the game
+        #    println("You crashed")
+        #else
+        obstacles_replaced = 0
+        while obstacles_replaced < w.num_obstacle
+            old_pos = w.obs_pos_array[obstacles_replaced+1]
+            next_pos = CartesianIndex(old_pos[1]+rand([-1,1]),old_pos[2]+rand([-1,1]))
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:59]: The same as above. Use a rng here.

File: src/envs/dynamicobstacles.jl [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                flag = 0
+                for i = obstacles_replaced:0
+                    if (i < obstacles_replaced) && (w.obs_pos_array[i+1]==next_pos)
+                        flag = 1
+                        break
+                    end
+                end
+                if flag == 1
+                    continue
+                    println("obstacles are overlapping")
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:72]: `@debug` would be more here suitable here compared to println ;)

File: src/envs/dynamicobstacles.jl [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        #    println("You crashed")
+        #else
+        obstacles_replaced = 0
+        while obstacles_replaced < w.num_obstacle
+            old_pos = w.obs_pos_array[obstacles_replaced+1]
+            next_pos = CartesianIndex(old_pos[1]+rand([-1,1]),old_pos[2]+rand([-1,1])) 
+            if w.world[WALL, next_pos] 
+                continue
+            else
+                flag = 0
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:63]: Either rename the flag to a more meaningful name or add a comment here to increase the readability

File: src/envs/dynamicobstacles.jl [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                for i = obstacles_replaced:0
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:64]: I guess the for loop will never enter the inner expression here. You may verify it by `collect(3:0)`

File: src/envs/dynamicobstacles.jl [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                    if (i < obstacles_replaced) && (w.obs_pos_array[i+1]==next_pos)
```
Comment (findmyway) [inline src/envs/dynamicobstacles.jl:65]: Could you explain the logic here a little bit? I think in some extreme cases the while loop will never end. For example, an obstacle surrounded by 4 other obstacles.

**Review by findmyway — CHANGES_REQUESTED** [review]
Awesome work! Added some comments to help improve the implementation further.

File: src/envs/dynamicobstacles.jl (hunk 2 above)
Comment (RajGhugare19) [reply src/envs/dynamicobstacles.jl:23]: Okay, i'll remove this line

File: src/envs/doorkey.jl (hunk 1 above)
Comment (RajGhugare19) [reply src/envs/doorkey.jl:18]: Yeah I will remove it

File: src/envs/dynamicobstacles.jl (hunk 9 above)
Comment (RajGhugare19) [reply src/envs/dynamicobstacles.jl:65]: Yeah I am looking into this, maybe  if instead of rand([1,-1]), If I do rand([-1,0,1]) would solve this

Comment (findmyway) [reply src/envs/dynamicobstacles.jl:65]: Sounds like a clever trick. 👍

**Comment by RajGhugare19** [comment]
@findmyway  I think it should work now, please tell me if something is wrong, I had messed up the for loop :p (new to julia)

**Pull request closed without merging**